Pod level in place pod resize - alpha

ndixita opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind api-change
#### What this PR does / why we need it:
This PR adds support for IPPR + Pod Level resources to allow in place resizing of pod-level resources in alpha stage.
#### Which issue(s) this PR is related to:
Issue: https://github.com/kubernetes/enhancements/issues/5419

Reference KEPs when applicable in addition to specific issues.

KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/5419-pod-level-resources-in-place-resize/README.md
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md

#### Does this PR introduce a user-facing change?

```release-note
This change allows In Place Resize of Pod Level Resources 
- Add Resources in PodStatus to capture resources set at pod-level cgroup
- Add AllocatedResources in PodStatus to capture resources requested in the PodSpec
```

ndixita committed:
API changes for Pod Level IPPR related PodStatus fields

ndixita commented:
/test all

ndixita commented:
/test pull-kubernetes-unit

ndixita commented:
/test pull-kubernetes-e2e-gce

ndixita commented:
/test all

ndixita commented:
/test all

ndixita commented:
/retest

ndixita commented:
/test all

ndixita commented:
/test pull-kubernetes-unit


<br class="Apple-interchange-newline">

tallclair reviewed (commented):
Good start, we need to figure out some nuances of the API though.

I assume you're still working on the changes to update kuberuntime_manager?

tallclair commented:
/label api-review

tallclair commented:
/milestone v1.35

ndixita committed:
Adding Resources and AllocatedResoures fields to the list of expected fields in PodStatus in admission test

ndixita committed:
Generated files from API changes

ndixita committed:
Adding InPlacePodLevelResourcesVerticalScaling feature gate

ndixita committed:
Modifying validation logic and pod strategy for PodResize to accommodate for Pod Level Resources resize

ndixita committed:
Scheduler changes to support pod level resources in place resize

ndixita committed:
Kubelet changes to support IPPR + PodLevelResources

Signed-off-by: ndixita <ndixita@google.com>

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -4350,6 +4350,27 @@ type PodStatus struct {
 	ResourceClaimStatuses []PodResourceClaimStatus
+
+	// AllocatedResources is the total requests allocated for this pod by the node.
+	// Kubelet sets this to the accepted requests when a pod (or resize) is admitted.
```
I think the API server should set this before the pod is bound (or maybe just on create), so let's drop this implementation detail.

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -4350,6 +4350,27 @@ type PodStatus struct {
+	// Pod-level limit aggregation is only performed, and is meaningful only, when all
+	// containers have defined limits.
+	// +featureGate=InPlacePodVerticalScaling
+	// +featureGate=PodLevelResources
```
Just set the `InPlacePodLevelResourcesVerticalScaling` feature here, and use the new feature gate dependencies to depend on the other 2.

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5088,6 +5088,27 @@ type PodStatus struct {
+	// +featureGate=InPlacePodVerticalScaling
+	// +featureGate=PodLevelResources
+	// +optional
+	Resources *ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,19,opt,name=resources"`
```
(for API discussion) it was maybe a mistake that we reused this type for container resources, since the `Claims` field is unused. Should we create a new type to use here, or stick with ResourceRequirements for consistency with the container level fields?

tallclair commented on pkg/api/pod/util.go:
```diff
@@ -884,6 +885,12 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
 		podStatus = &api.PodStatus{}
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) && !inPlacePodVerticalScalingInUse(oldPodSpec) {
```
I think you need to add a new in-use function for this check, otherwise you could set these fields without the feature gate by setting the container-level fields (which will always be set automatically) and then setting the pod-level fields.

tallclair commented on pkg/api/pod/util.go:
```diff
@@ -895,6 +902,9 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
 		dropResourcesField(podStatus.EphemeralContainerStatuses)
 
+		// Drop pod-level Resources fields from PodStatus
+		podStatus.Resources = nil
```
nit: you don't need to drop the resources here. With FG dependencies, this is redundant with the previous block.

ndixita commented:
/test all

natasha41575 commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -201,6 +201,10 @@ func SetDefaults_Pod(obj *v1.Pod) {
 		defaultPodRequests(obj)
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) && utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
```
Per https://github.com/kubernetes/kubernetes/pull/133697 and https://github.com/kubernetes/kubernetes/pull/133912, I believe the correct / recommended way to do this now is to declare `InPlacePodLevelResourcesVerticalScaling` as having a dependency on `PodLevelResources`, and then here we only need to check `InPlacePodLevelResourcesVerticalScaling`.

@tallclair can confirm though 

Also here's the email ICYMI: https://groups.google.com/a/kubernetes.io/g/dev/c/OpUWIVR9DHY/m/1rFE2eaqAQAJ?utm_medium=email&utm_source=footer

natasha41575 commented on pkg/api/pod/util.go:
```diff
@@ -1003,6 +1009,9 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
 		dropAllocatedResourcesField(podStatus.EphemeralContainerStatuses)
+
+		/// Drop pod-level AllocatedResources field
+		podStatus.AllocatedResources = nil
```
does this need to be here? `AllocatedResources` is cleared in the code block above (L986-990)

We're about to promote `InPlacePodVerticalScaling` to GA so this entire code block (L992-L1015) will become moot

liggitt replied:
yeah, the above line should be sufficient

natasha41575 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6117,6 +6107,41 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+
+		if !opts.PodLevelResourcesEnabled {
+			return field.ErrorList{field.Forbidden(field.NewPath(""), "pods with pod-level resources is forbidden when PodLevelResources feature gate is disabled")}
+		}
```
I think this will no longer be needed after you declare the feature gate dependency

natasha41575 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6219,33 +6243,116 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
-			mungedResourceList = resourceList.DeepCopy()
+// isPodLevelResourcesResizeRequestSupported checks whether the pod supports resize requests
+// by verifying if the pod is running and has pod-level resources populated.
+func isPodLevelResourcesResizeRequestSupported(pod core.Pod) bool {
```
The analogous function for InPlacePodVerticalScaling has a TODO to remove it after GA+3: https://github.com/kubernetes/kubernetes/blob/96867d8401409db5fc78dccbc43fb15706890427/pkg/apis/core/validation/validation.go#L6250-L6256

I suggest to leave a similar TODO comment here that we can remove it after GA+3

natasha41575 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6219,33 +6243,116 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+	return allErrs
+}
+
+func validateMemoryResize(memRestartPolicy core.ResourceResizeRestartPolicy, newPodSpec, oldPodSpec core.PodSpec) field.ErrorList {
```
We allow memory limit decrease for container-level IPPR now. I don't see it written in the KEP that we will forbid this for PLR IPPR, so I think we want to support it for parity unless there is some technical constraint that I am not aware of? 

Non-blocking for alpha I think, but at minimum we should add a TODO and revisit it in beta

natasha41575 commented on pkg/kubelet/allocation/allocation_manager.go:
```diff
@@ -437,6 +438,16 @@ func updatePodFromAllocation(pod *v1.Pod, allocated state.PodResourceInfo) (*v1.
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) && utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources) {
+		pAlloc := allocated.PodLevelResources
+		if !apiequality.Semantic.DeepEqual(pod.Spec.Resources, pAlloc) {
+			if !updated {
```
`updated` is always false at the line, isn't it? I don't think we need this `if`

natasha41575 commented on pkg/kubelet/allocation/features_linux.go:
```diff
@@ -35,3 +35,7 @@ func IsInPlacePodVerticalScalingAllowed(pod *v1.Pod) (allowed bool, msg, reason
 	return true, "", ""
 }
+
+func IsInPlacePodLevelResourcesVerticalScalingAllowed(pod *v1.Pod) (allowed bool) {
```
I'm a little surprised not to see any changes in `handlePodResourcesResize`. Does this need to be called from there as part of the feasibility checks?

natasha41575 reviewed (commented):
I have a few more open questions but I think they are likely to be answered / addressed naturally after you get unit and e2e tests passing, so I'm holding them for now

liggitt commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -528,3 +532,14 @@ func defaultHugePagePodLimits(pod *v1.Pod) {
+		return
+	}
+
+	obj.Status.AllocatedResources = make(v1.ResourceList)
```
if we keep doing this as generic API defaulting (rather than moving to PrepareForCreate/PrepareForUpdate), this needs to be conditional on obj.Status.AllocatedResources being nil or this isn't defaulting, it's always unconditionally stomping

liggitt commented on pkg/registry/core/pod/strategy.go:
need to mirror the population of the field in status in `PrepareForCreate` that we're doing in defaulting, if we want it to be persisted into etcd on create

```
	pod.Status = api.PodStatus{
		Phase:    api.PodPending,
		QOSClass: qos.GetPodQOS(pod),
	}
	if feature-gate-enabled { pod.Status.AllocationResources = ... }
```

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6117,6 +6107,41 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+		}
+
+		allErrs = append(allErrs, validatePodLevelResourcesResize(newPod, oldPod, specPath, opts)...)
+		dropCPUMemoryResourcesFromPodLevelResources(originalCPUMemPodSpec.Resources, oldPod.Spec.Resources)
```
it's hard to tell if this is mutating stuff under oldPod ... is there a way to make this more obviously non-mutating, extracting and passing in the oldPod cpu/memory bits instead of the whole parent object that could be mutated?

ndixita commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -528,3 +532,14 @@ func defaultHugePagePodLimits(pod *v1.Pod) {
 	}
 }
+
+func defaultPodStatusAllocatedResources(obj *v1.Pod) {
```
The initial plan was to populate AllocatedResources equal to desired resources specified in the pod spec for other components (autoscaler, etc) to determine the total requested resources before a pod is scheduled. 

But after the recent discussion with API folks, this field might not have a reliable value if set in control plane that support this feature, and then the pod is handled by a node without this feature.  Jordan's [comment](https://github.com/kubernetes/kubernetes/pull/132919/commits/ef89cb07e3837f948ab15e3d104bfcde8f8c6499#r2456402778) for more details

ndixita commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -38,10 +39,16 @@ type PodResourcesOptions struct {
+	// PodStatus should be considered when evaluating the pod resources.
+	// This MUST be false if the InPlacePodLevelResourcesVerticalScaling
+	// feature is not enabled.
+	UsePodStatusResources bool
```
Well, as much as I agree with this change of associating with the feature gate, I think we might not get approvals for removing a public variable once it is exposed in the component helper library. I will confirm this. And we are not allowed to drop the fields once they are made public, I wouldn't want to tie the change with feature gate name.

liggitt reviewed (commented):
just a few cleanup bits in the validation code, API bits lgtm otherwise

natasha41575 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1943,6 +1943,11 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+		// TODO(ndixita) move to status manager and set it equal to pod requests
+		pod.Status.AllocatedResources = pod.Spec.Resources.Requests
```
for container resize, this is done here: https://github.com/kubernetes/kubernetes/blob/9630ab9581afbac9835d53f9e620a1240a1d2d91/pkg/kubelet/kubelet_pods.go#L2411

your todo to move this to status manager is throwing me off, was there a reason to do it there rather than in kubelet_pods.go like we do for container resize?

ndixita commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

tallclair commented:
/assign @tallclair @natasha41575 @liggitt

tallclair commented on pkg/api/pod/util_test.go:
```diff
@@ -6264,3 +6264,145 @@ func TestHasUserNamespacesWithVolumeDevices(t *testing.T) {
+					if newPod == nil {
+						continue
+					}
+					t.Run(fmt.Sprintf("old pod %v, new pod %v", oldPodInfo.description, newPodInfo.description), func(t *testing.T) {
```
nit: this is going to result in absurdly long test names. Consider giving the pods shorter identifier names.

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -4550,6 +4550,20 @@ type PodStatus struct {
+
+	// Resources represents the compute resource requests and limits that have been
+	// applied at the pod level if pod-level requests or limits are set in
+	// PodSpec.Resources
```
```suggestion
	// applied at the pod level
```

Didn't we decide that this would always reflect the values used in the pod-level cgroups? So it would still be set if pod-level limits were applied?

ndixita commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features

dom4ha commented:
/cc @macsko @sanposhiho

sanposhiho commented on pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:
any new tests for changes? I only see new tests for QHint

macsko replied:
Maybe worth adding at least:
- an integration test case for QHint in `test/integration/scheduler/queueing/queue.go`
- test cases in `TestCalculatePodResourcesWithResize`

ndixita committed:
e2e tests

Signed-off-by: ndixita <ndixita@google.com>

natasha41575 commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -173,8 +197,23 @@ func TestUpdatePodFromAllocation(t *testing.T) {
 		expectUpdate: true,
 		expectPod:    resizedPod,
+	}, {
+		name: "with resized pod-level resource allocation",
```
Can we also have a cases with the feature gate off, and a case with just pod level resources (no container resources specified)?

natasha41575 commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -173,8 +197,23 @@ func TestUpdatePodFromAllocation(t *testing.T) {
+		expectPod:    resizedPodWithPodLevelResources,
 	}}
 
+	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodLevelResourcesVerticalScaling, true)
```
General comment: This file needs more updates; we need to modify TestRetryPendingResizes, TestAllocationManagerAddPod, TestIsResizeIncreasingRequests, and TestSortPendingResizes to include coverage for pod-level resources

natasha41575 commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -74,10 +75,15 @@ func offsetMemory(index int64, value string) string {
+				e2eskipper.Skipf("Skip pod level resource resize")
+			}
+			if !usePodLevelResources && (desiredPodCPU != "" || desiredPodMem != "") {
+				e2eskipper.Skipf("Skip pod level resource resize")
```
```suggestion
			if !usePodLevelResources && (desiredPodCPU != "" || desiredPodMem != "") {
				e2eskipper.Skipf("Skip pod level resource resize")
```

natasha41575 commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2106,9 +2107,98 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
+		preserveOldResourcesValue(v1.ResourceMemory, oldPodStatus.Resources.Limits, resources.Limits)
+	}
+
+	return resources
```
A lot of this function seems to be duplicating the equivalent container function here: https://github.com/ndixita/kubernetes/blob/175c254e6d3df3bc4a09408252ecf137c161ac98/pkg/kubelet/kubelet_pods.go#L2274-L2333

Is it possible to extract to a helper?

dom4ha commented on pkg/scheduler/framework/plugins/feature/feature.go:
```diff
@@ -25,26 +25,27 @@ import (
+	EnableNodeInclusionPolicyInPodTopologySpread  bool
+	EnableMatchLabelKeysInPodTopologySpread       bool
+	EnableInPlacePodVerticalScaling               bool
+	EnableInPlacePodLevelResourcesVerticalScaling bool
```
Please set the value in `NewSchedulerFeaturesFromGates` below as well.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6219,33 +6236,129 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+
+func dropCPUMemoryUpdates(resourceList, oldResourceList core.ResourceList) core.ResourceList {
+	if oldResourceList == nil {
+		return nil
```
Wouldn't this drop non-cpu/mem changes too?

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6219,33 +6236,129 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+}
+
+// dropCPUMemoryResourcesFromContainer deletes the cpu and memory resources from the container, and copies them from the old pod container resources if present.
+func dropCPUMemoryResourcesFromContainer(container *core.Container, oldPodSpecContainer *core.Container) {
```
nit: delete this function: just use `dropCPUMemoryResourcesFromPodLevelResources` (renamed) for containers too. E.g. `dropCPUMemoryResources(&container.Resources, &oldContainer.Resources)`

liggitt replied:
actually, because of the "Set the entire Resources block to nil if two conditions are met" block in that method, let's defer collapsing to a single method to a follow up

tallclair commented on pkg/kubelet/allocation/features_linux.go:
```diff
@@ -35,3 +35,10 @@ func IsInPlacePodVerticalScalingAllowed(pod *v1.Pod) (allowed bool, msg, reason
+
+func IsInPlacePodLevelResourcesVerticalScalingAllowed(pod *v1.Pod) (allowed bool, msg, reason string) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+		return false, "InPlacePodLevelResourcesVerticalScaling is disabled", "feature_gate_off"
```
nit: the reason is missing the context
```suggestion
		return false, "InPlacePodLevelResourcesVerticalScaling is disabled", "plr_feature_gate_off"
```

tallclair commented on pkg/kubelet/cm/helpers_linux.go:
```diff
@@ -28,7 +29,9 @@ import (
 
-	"k8s.io/component-helpers/resource"
+	"k8s.io/apimachinery/pkg/api/resource"
+
```
nit: remove this blank line, and the one above.

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2106,9 +2107,98 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
+	// the code needs to persist that value if it is non-nil.
+	preserveOldResourcesValue(v1.ResourceMemory, oldPodStatus.Resources.Requests, resources.Requests)
+
+	if cpuLimit != nil {
```
Same here - if the cgroup limit is "max", what value do we get here?

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -581,6 +581,14 @@ type podActions struct {
 }
 
+// podLevelResources holds the set of resources applicable to the running pod
+type podLevelResources struct {
```
This doesn't need to be a separate type from container resources. Just reuse `containerResources` for now, and leave a TODO to rename it to something more generic.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -794,6 +817,14 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(ctx context.Context, pod *
 		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+			if err = m.actuatedState.SetPodLevelResources(pod.UID, pod.Spec.Resources); err != nil {
```
You also need to initialize the actuated state when the pod is first created. Here is where we currently initialize the pod cgroup: https://github.com/kubernetes/kubernetes/blob/3ec72e2e973664a0690bc7122c9cd291702b5f36/pkg/kubelet/kubelet.go#L2064

I guess that would be the best place to call it?

tallclair commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -125,6 +132,9 @@ func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
 				reqs[resourceName] = quantity
+				if opts.UsePodStatusResources {
+					reqs[resourceName] = determinePodResourceReqs(pod, resourceName)
+				}
```
This comment still needs to be addressed.

ndixita commented:
/retest

ndixita commented:
/retest

natasha41575 commented on pkg/kubelet/kubelet.go:
```diff
@@ -2780,14 +2780,49 @@ func (kl *Kubelet) HandlePodUpdates(pods []*v1.Pod) {
+	hasResize := false
+	hasResize = recordPodLevelResourceResizeOperations(oldPod, newPod)
+	hasResize = recordContainerResizeOperations(oldPod, newPod)
+	return hasResize
```
You are unconditionally overwriting the pod-level hasResize with the container-level hasResize, making the first one ineffectual

```suggestion
	return recordPodLevelResourceResizeOperations(oldPod, newPod) || recordContainerResizeOperations(oldPod, newPod)
```

ndixita commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2106,9 +2107,98 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
+		resources.Limits = make(v1.ResourceList)
+	}
+
+	if cpuRequest != nil {
```
always comes up as nil as CPURequestsFromConfig method returns non-nil value only if cpu shares is greater than 0

liggitt reviewed (commented):
I reviewed all the API bits except validatePodLevelResourcesResize and dropCPUMemoryResources* where it looked @tallclair had a couple outstanding comments

once those are addressed I'll sweep validation.go and then can tag for API bits

dom4ha commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -376,11 +376,12 @@ func (f *Fit) isSchedulableAfterPodScaleDown(targetPod, originalPod, modifiedPod
 
 	// check whether the resource request of the modified pod is less than the original pod.
-	podRequests := resource.PodRequests(targetPod, resource.PodResourcesOptions{UseStatusResources: f.enableInPlacePodVerticalScaling})
+	podRequests := resource.PodRequests(targetPod, resource.PodResourcesOptions{UseStatusResources: f.enableInPlacePodVerticalScaling, InPlacePodLevelResourcesVerticalScalingEnabled: f.enableInPlacePodLevelResourcesVerticalScaling})
```
Use opts here as well.

ndixita commented:
/test pull-kubernetes-e2e-kind-alpha-beta-enabled-conformance

ndixita commented on pkg/kubelet/kubelet.go:
```diff
@@ -2066,6 +2066,10 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
+
+				if err = kl.containerRuntime.UpdateActuatedPodLevelResources(pod); err != nil {
+					return false, fmt.Errorf("failed to update the state of pod-level resources for the pod %v : %v", pod.UID, err)
+				}
```
@tallclair Does this look right?

tallclair commented on pkg/scheduler/framework/types_test.go:
```diff
@@ -1623,6 +1623,8 @@ func TestPodInfoCalculateResources(t *testing.T) {
 
 func TestCalculatePodResourcesWithResize(t *testing.T) {
 	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodVerticalScaling, true)
+	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodLevelResourcesVerticalScaling, true)
```
nit: Use the bulk method
```suggestion
	featuregatetesting.SetFeatureGatesDuringTest(t, utilfeature.DefaultFeatureGate, featruegatetesting.FeatureOverrides{
		features.InPlacePodVerticalScaling: true,
		features.InPlacePodLevelResourcesVerticalScaling: true,
	})
```

tallclair commented on pkg/kubelet/allocation/state/state_mem.go:
```diff
@@ -59,6 +59,23 @@ func (s *stateMemory) GetContainerResources(podUID types.UID, containerName stri
+		return nil, ok
+	}
+
+	if pr.PodLevelResources != nil {
```
Shouldn't this be `if pr.PodLevelResources == nil`? This seems like a pretty fundamental problem - unless I'm missing something, it seems concerning that tests are passing with this error.

ndixita commented:
/test all

ndixita commented:
/retest

ndixita commented:
/test pull-kubernetes-node-kubelet-serial-podlevelresources

liggitt commented on pkg/apis/core/validation/validation.go:
note to self, this file was clearer to review with whitespace ignored to make the dropCPUMemoryUpdates extraction more obvious

https://github.com/kubernetes/kubernetes/pull/132919/files?w=1

natasha41575 commented:
@ndixita Looks like all the resize e2e tests are being skipped?

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/132919/pull-kubernetes-node-kubelet-inplace-pod-level-resources-resize/1986540156575813632

ndixita commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6260,33 +6277,110 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+// dropCPUMemoryResourcesFromPodLevelResources deletes the cpu and memory resources
+// from the `resources` field, and copies them from old Pod spec's resources field
+// if present.
+func dropCPUMemoryResourcesFromPodLevelResources(resources *core.ResourceRequirements, oldPodResources *core.ResourceRequirements) *core.ResourceRequirements {
```
the control doesn't come to this method if resources is nil. But anyway added the check for safety

ndixita committed:
Kubelet changes and unit tests

Signed-off-by: ndixita <ndixita@google.com>

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6265,33 +6282,116 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
+	}
+
+	if oldPod.Spec.Resources != nil {
+		if oldPod.Spec.Resources.Requests != nil {
```
non-blocking: the `oldPod.Spec.Resources.Requests != nil` check isn't necessary, could call resourcesRemoved without it to simplify

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6265,33 +6282,116 @@ func validatePodResizeContainerOrdering(newPod, oldPod *core.Pod, specPath *fiel
-		if cpu, found := oldResourceList[core.ResourceCPU]; found {
-			mungedResourceList[core.ResourceCPU] = cpu
+
+		if oldPod.Spec.Resources.Limits != nil {
```
non-blocking: the `oldPod.Spec.Resources.Limits != nil` check isn't necessary, could call resourcesRemoved without it to simplify

liggitt reviewed (commented):
API bits lgtm, @tallclair has lgtm on test coverage and impl

once those bits are approved I can add approval on API packages

pravk03 committed:
Add InPlacePodLevelResourcesVerticalScaling declared feature.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
Add pod level resources test cases to `TestComputePodActionsForPodResize`. This is the test that should have caught the error in the actuated state.

ndixita commented:
/test pull-kubernetes-node-kubelet-inplace-pod-level-resources-resize

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -1904,6 +2037,48 @@ func (m *kubeGenericRuntimeManager) IsPodResizeInProgress(allocatedPod *v1.Pod,
+	actuatedPodResources, _ := m.actuatedState.GetPodLevelResources(allocatedPod.UID)
+	allocatedPodResources := allocatedPod.Spec.Resources
+
+	return cpuMemoryResourcesEqual(actuatedPodResources, allocatedPodResources)
```
This is inverted. 
```suggestion
	return !cpuMemoryResourcesEqual(actuatedPodResources, allocatedPodResources)
```

Here's unit test coverage: https://github.com/tallclair/kubernetes/commit/e21d01fdba159b414e3fcf40f8692355c9569e1b

ndixita replied:
looking into it

natasha41575 commented:
Btw, not a blocker for alpha, but for beta we need to add the pod-level resources to the resize events. It looks a lil funny rn when resizing pod-level resources:

```
  Normal  ResizeStarted    15s   kubelet            Pod resize started: {"containers":[{"name":"nginx","resources":{}}],"generation":2}
  Normal  ResizeCompleted  15s   kubelet            Pod resize completed: {"containers":[{"name":"nginx","resources":{}}],"generation":2}
```

ndixita commented:
> I added some E2E coverage for resizing a pod with only pod-level resources: [tallclair@67790e5](https://github.com/tallclair/kubernetes/commit/67790e5a51ecb44b064db2b93d279e531ba36cc5)
> 
> This uncovered an error: on create, we set the container limits to the pod limits if the container limits are unset, but we don't resize the container when resizing the pod limits. Changes needed:
> 
> 1. Factor in pod-level resources when computing container resize actions: https://github.com/kubernetes/kubernetes/blob/173523550ab112ca956d74238e050a0a2ce2a5e3/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L652
> 2. Factor in pod-level resources when setting container actuated resources: https://github.com/tallclair/kubernetes/blob/67790e5a51ecb44b064db2b93d279e531ba36cc5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L264

Agree with 1.

2. Requires some more thinking... If we update just the actuated resources, the allocated resources won't be equal to actuated resources and IsPodResizeInProgress will always return true. I have added a check in IsPodResizeInProgress as well for now. I think we could rather have common defaulting logic somewhere (will try to think about this for next release) before resources are set in allocated and actuated states, and in cgroups.

tallclair commented:
> If we update just the actuated resources, the allocated resources won't be equal to actuated resources and IsPodResizeInProgress will always return true. I am also adding the logic for allocated resources.

Hmm, good point. I think we still need to track this though. Consider this case:
1. Decreasing PLR limits, 2 containers without resources
2. Container 1 is resized successfully
3. Container 2 fails to resize (e.g. memory limit decrease below usage)
4. Since there were failures resizing the containers, the pod cgroups won't be updated, and pod actuated resources won't be set.

If we just use pod actuated resources as a proxy for container actuated resources in this case, then container 1 will be resized on every pod sync. This isn't a huge issue if it doesn't need a restart (resize should be idempotent), but it becomes a big problem if the container has a `RestartRequired` resize policy.

If we include implicit PLR resources in container-level actuated resources, I think there are 2 options for resolving the difference between actuated & allocated resources:
1. Include the computed PLR in allocated resources - this mostly seems fine, but it will leak into the API through the allocatedResources status field. We could sort of avoid this by checking the container spec when generating the api status, but that won't work if we add the ability to remove resources (https://github.com/kubernetes/kubernetes/pull/127143).
2. Add the implicit PLRs into the container-allocated resources before comparing allocated & actuated here: https://github.com/kubernetes/kubernetes/blob/ef274e869c3ea3d4042fada38a56221283a42362/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1898

I think option 2 seems better.

tallclair reviewed (commented):
I ran locally and confirmed the tests are failing. Please try to rerun the tests before pushing changes to github. It makes it more difficult to review if I don't know the current state of the PR.

natasha41575 commented on pkg/kubelet/allocation/allocation_manager.go:
```diff
@@ -440,6 +440,15 @@ func updatePodFromAllocation(pod *v1.Pod, allocated state.PodResourceInfo) (*v1.
 	}
 
 	updated := false
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
```
Still hoping this comment can be addressed: https://github.com/kubernetes/kubernetes/pull/132919#discussion_r2487604622

ndixita commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -608,8 +620,30 @@ func containerSucceeded(c *v1.Container, podStatus *kubecontainer.PodStatus) boo
+	resources.memoryLimit = containerLimits.Memory().Value()
+	resources.cpuLimit = containerLimits.Cpu().MilliValue()
+	return resources
+}
```
@natasha41575 here

KevinTMtz committed:
Unit tests for allocation manager

ndixita committed:
Unit test for AddPod with PLR resources

Signed-off-by: ndixita <ndixita@google.com>

ndixita committed:
verioned_feature_list update

ndixita committed:
Test fixes

Signed-off-by: ndixita <ndixita@google.com>

natasha41575 commented on pkg/kubelet/allocation/state/state.go:
```diff
@@ -51,11 +53,13 @@ type Reader interface {
 	GetContainerResources(podUID types.UID, containerName string) (v1.ResourceRequirements, bool)
 	GetPodResourceInfoMap() PodResourceInfoMap
 	GetPodResourceInfo(podUID types.UID) (PodResourceInfo, bool)
+	GetPodLevelResources(podUID types.UID) (*v1.ResourceRequirements, bool)
```
(tiny nit) the analogous `GetContainerResources` returns `(v1.ResourceRequirements, bool)`, should we keep it consistent and have the same return values?

natasha41575 commented on pkg/kubelet/allocation/state/state.go:
```diff
@@ -51,11 +53,13 @@ type Reader interface {
 type writer interface {
 	SetContainerResources(podUID types.UID, containerName string, resources v1.ResourceRequirements) error
 	SetPodResourceInfo(podUID types.UID, resourceInfo PodResourceInfo) error
+	SetPodLevelResources(podUID types.UID, alloc *v1.ResourceRequirements) error
```
similar comment here

natasha41575 commented on pkg/kubelet/allocation/state/state_mem.go:
```diff
@@ -77,17 +90,41 @@ func (s *stateMemory) SetContainerResources(podUID types.UID, containerName stri
+
+	podInfo, ok := s.podResources[podUID]
+	if !ok {
+		podInfo.PodLevelResources = &v1.ResourceRequirements{}
```
does this line do anything? You are immediately overwriting it in the next line

natasha41575 commented on pkg/kubelet/allocation/state/state_mem.go:
```diff
@@ -77,17 +90,41 @@ func (s *stateMemory) SetContainerResources(podUID types.UID, containerName stri
+	podInfo, ok := s.podResources[podUID]
+	if !ok {
+		podInfo = PodResourceInfo{
 			ContainerResources: make(map[string]v1.ResourceRequirements),
```
Are these lines necessary? The returned podInfo from `podInfo, ok := s.podResources[podUID]` should be a `PodResourceInfo{}`, so L100-102 should cover the `!ok` case

natasha41575 commented on pkg/kubelet/allocation/allocation_manager.go:
```diff
@@ -559,6 +580,15 @@ func (m *manager) handlePodResourcesResize(pod *v1.Pod) (bool, error) {
 		return false, nil
 	}
 
+	if !apiequality.Semantic.DeepEqual(pod.Spec.Resources, allocatedPod.Spec.Resources) {
```
This check should be behind the PodLevelResources feature gate

natasha41575 commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -281,19 +377,23 @@ func TestRetryPendingResizes(t *testing.T) {
+		originalRequests             v1.ResourceList
+		newRequests                  v1.ResourceList
+		originalLimits               v1.ResourceList
+		newLimits                    v1.ResourceList
```
nit: rename these to `originalContainerRequests` etc

natasha41575 commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -548,11 +648,90 @@ func TestRetryPendingResizes(t *testing.T) {
 			expectPodSyncTriggered: "true",
 		},
+		{
+			name:                         "pod-level: Request CPU and memory increase - expect InProgress",
```
Thanks for adding these cases. They look good. I would like to request one or two cases with both pod level and container level resources

natasha41575 commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -548,11 +648,90 @@ func TestRetryPendingResizes(t *testing.T) {
 			t.Run(fmt.Sprintf("%s/sidecar=%t", tt.name, isSidecarContainer), func(t *testing.T) {
+				if tt.inPlacePodLevelResizeEnabled {
+					featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodLevelResourcesVerticalScaling, true)
+				}
```
Suggested change
      
    
    
      
          
            
            				if tt.inPlacePodLevelResizeEnabled {
          
          
            
            					featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodLevelResourcesVerticalScaling, true)
          
          
            
            				}
          
          
            
            				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodLevelResourcesVerticalScaling, tt.inPlacePodLevelResizeEnabled)

I see a few other places actually where the same comment applies, please update them all

natasha41575 commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -656,16 +689,21 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(ctx context.Context,
-	currentResources := containerResourcesFromRequirements(&actuatedResources)
+	var actuatedPodResources *v1.ResourceRequirements
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+		actuatedPodResources, _ = m.actuatedState.GetPodLevelResources(pod.UID)
```
do we need to check the return value of `found` here (similar to what we do with container resources above)

tallclair replied:
It's just used for logging, but probably good to know?

natasha41575 commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -1963,12 +2080,38 @@ func (m *kubeGenericRuntimeManager) ListPodSandboxMetrics(ctx context.Context) (
+		return true
+	}
+
+	return m.isPodLevelResourcesResizeInProgress(allocatedPod, podStatus)
```
optional nit, just a personal style preference


```suggestion
	return m.isContainerResourceResizeInProgress(allocatedPod, podStatus) || m.isPodLevelResourcesResizeInProgress(allocatedPod, podStatus)
```

natasha41575 commented on test/e2e/common/node/pod_resize.go:
```diff
@@ -795,7 +807,7 @@ func doPodResizeReadAndReplaceTests(f *framework.Framework) {
 //	   b) api-server in services doesn't start with --enable-admission-plugins=ResourceQuota
 //	      and is not possible to start it from TEST_ARGS
 //	Above tests in test/e2e/node/pod_resize.go
-var _ = SIGDescribe("Pod InPlace Resize Container", func() {
```
i.e. make sure there's no diff when you run ./hack/update-conformance.sh

natasha41575 reviewed (commented):
> The WiFi isn't working at the conference, so I won't be able to review this until I get back to my hotel room tonight.
> 
> @natasha41575 can you help review today? If you lgtm I'll try to just approve tonight.

I'm out of time today. I have a few outstanding comments here and also above but besides this I think this largely LGTM if those are addressed

I will also add a caveat that I did not get around to looking at the e2e tests super carefully after they were updated to make everything pass - so that still needs another eye on it.

natasha41575 commented on test/e2e/common/node/pod_level_resources_resize.go:
```diff
@@ -0,0 +1,732 @@
+/*
+Copyright 2024 The Kubernetes Authors.
```
2025

tallclair replied:
Better yet, just remove the year. It's not needed any more.

natasha41575 commented on test/e2e/node/pod_resize.go:
```diff
@@ -77,22 +77,22 @@ func doPodResizeResourceQuotaTests(f *framework.Framework) {
 
 	ginkgo.DescribeTable("pod-resize-resource-quota-test",
-		func(ctx context.Context, desiredContainers []podresize.ResizableContainerInfo, expectedContainers []podresize.ResizableContainerInfo, wantError string) {
+		func(ctx context.Context, desiredContainers []podresize.ResizableContainerInfo, expectedContainers []podresize.ResizableContainerInfo, podResources *v1.ResourceRequirements, wantError string) {
```
looks like `podResources` is nil in every test case, can we just remove the parameter?

ndixita replied:
Yes, need to cleanup. I ended up creating a separate test file altogether so this is N/A now.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -594,11 +595,21 @@ type podActions struct {
 func (p podActions) String() string {
-	return fmt.Sprintf("KillPod: %t, CreateSandbox: %t, UpdatePodResources: %t, Attempt: %d, InitContainersToStart: %v, ContainersToStart: %v, EphemeralContainersToStart: %v,ContainersToUpdate: %v, ContainersToKill: %v, ContainersToRemove: %v",
-		p.KillPod, p.CreateSandbox, p.UpdatePodResources, p.Attempt, p.InitContainersToStart, p.ContainersToStart, p.EphemeralContainersToStart, p.ContainersToUpdate, p.ContainersToKill, p.ContainersToReset)
+	return fmt.Sprintf("KillPod: %t, CreateSandbox: %t, UpdatePodResources: %t, UpdatePodLevelResources: %t, Attempt: %d, InitContainersToStart: %v, ContainersToStart: %v, EphemeralContainersToStart: %v,ContainersToUpdate: %v, ContainersToKill: %v, ContainersToRemove: %v",
```
eh, I think it's ok. This is just for debugging purposes.

tallclair commented on pkg/kubelet/allocation/allocation_manager_test.go:
```diff
@@ -126,16 +126,45 @@ func TestUpdatePodFromAllocation(t *testing.T) {
+
+	podWithoutContainerResources := pod.DeepCopy()
+	podWithoutContainerResources.Spec.Containers[0].Resources = v1.ResourceRequirements{}
+	podWithoutContainerResources.Spec.InitContainers[0].Resources = v1.ResourceRequirements{}
```
I think the pod has 2 regular containers and 2 init containers. Also, did you mean for this pod to have pod-level resources?

tallclair commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +60,52 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+// The implementation of CPU/Memory limit calculation is duplicated here and in
+// pkg/kubelet/kuberuntime/kuberuntime_container_linux.go (getCPULimits/getMemoryLimits).
+// Refactor into a shared utility function.
+func GetLimits(res *ResourceOpts) v1.ResourceList {
```
nit: why the ResourceOpts struct, rather than just passing 2 parameters?

ndixita replied:
While testing I swapped the two params :| so for clarity or I could pass container resources as the v1.ResourceRequirements and pod resources as *v1.ResourceRequirements.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -794,8 +863,16 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(ctx context.Context, pod *
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+			if err = updateActuatedPodLevelResources(rName); err != nil {
+				logger.Error(err, "Failed to update pod-level actuated resources", "resource", rName, "pod", klog.KObj(pod))
```
Still need to address this.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -1963,12 +2080,38 @@ func (m *kubeGenericRuntimeManager) ListPodSandboxMetrics(ctx context.Context) (
+		return nil
+	}
+
+	if actuatedPod.Spec.Resources == nil {
```
nit: this and the nil check below seem unnecessary.

tallclair commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -2123,9 +2124,124 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodLevelResourcesVerticalScaling) {
+		apiPodStatus.Resources = kl.convertToAPIPodLevelResourcesStatus(pod, oldPodStatus)
+		opts := resourcehelper.PodResourcesOptions{
+			SkipPodLevelResources: !utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources),
```
This should exclude overhead, right?

tallclair commented on staging/src/k8s.io/component-helpers/resource/helpers.go:
```diff
@@ -148,9 +153,25 @@ func PodRequests(pod *v1.Pod, opts PodResourcesOptions) v1.ResourceList {
 	if !opts.SkipPodLevelResources && IsPodLevelRequestsSet(pod) {
+
+		var effectiveReqs v1.ResourceList
+		if opts.InPlacePodLevelResourcesVerticalScalingEnabled && opts.UseStatusResources {
```
These changes should have unit test coverage.

tallclair reviewed (commented):
Were you going to pull https://github.com/kubernetes/kubernetes/pull/135203 into this PR?

I don't think any of the outstanding feedback is blocking for alpha. I'll take another look in a couple hours. In the meantime, try to track down the missing approvals.

@liggitt I think this is ready for your approval.

/approve

tallclair commented on test/e2e/common/node/pod_level_resources_resize.go:
```diff
@@ -0,0 +1,710 @@
+		ginkgo.By("Patching pod with a slightly lowered memory limit")
+		viableLoweredLimit := []podresize.ResizableContainerInfo{{
+			Name: "c1",
+		}}
```
nit: remove this. there shouldn't be any container-level resizes, so just reuse the original `containers`

ndixita committed:
Test fixes

Signed-off-by: ndixita <ndixita@google.com>

ndixita commented:
/test pull-kubernetes-unit-windows-master

ndixita commented:
/retest

ndixita committed:
Enabling NodeDeclaredFeatures in unit tests

Signed-off-by: ndixita <ndixita@google.com>

ndixita commented:
/retest

ndixita commented:
/test pull-kubernetes-e2e-gce

tallclair commented:
/lgtm

Some follow-up to do, but I think this is good enough for alpha.
